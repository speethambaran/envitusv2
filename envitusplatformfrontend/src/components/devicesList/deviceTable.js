import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Container, Modal, Button, Row, Col, Form } from 'react-bootstrap';
import DeviceEditForm from '../../sections/deviceMngt/deviceEditForm';
import { MdAddCircleOutline, MdDelete, MdRemoveCircleOutline } from "react-icons/md";
import { FiFilter } from "react-icons/fi";
import { deleteDevice, getDeviceCount, getAllDevices, updateDeviceData } from '../../services/deviceApi';
import LiveDataTable from '../liveDataTable/liveDataTable';
import DarkOverlays from '../overlays/darkOverlays';
import Paginations from '../paginations/paginations';
import {toastr} from 'react-redux-toastr';
import Loader from 'react-loader-spinner';
import './deviceTable.scss';

export class DeviceTable extends Component {
    constructor(props) {
        super(props);
        this.state = {
            offset: {act: 0, deact: 0},
            limit: 10,
            table: {activated: '', deactivated: '', actCnt: 0, deactCnt: 0},
            showEditModal: false,
            updateType: '',
            updateDeviceId: '',
            showDeleteModal: false,
            deleteDeviceId: '',
            modal: {head: '', content: {body: '', buttons: '', submitLink: ''}},
            apiCall: {restore: '', permanent: ''},
            deactivationReason: '',
            filter: {type: ['deviceId', 'zone', 'city', 'subType'], option: {}},
            filterCurrent: {text: '', select: ''},
        }
        this.showAddModal = this.showAddModal.bind(this);
        this.showFilterModal = this.showFilterModal.bind(this);
        this.showUpdateModal = this.showUpdateModal.bind(this);
        this.closeEditModal = this.closeEditModal.bind(this);
        this.showTempDeleteModal = this.showTempDeleteModal.bind(this);
        this.showDeleteModal = this.showDeleteModal.bind(this);
        this.showRestoreModal = this.showRestoreModal.bind(this);
        this.viewDevDetails = this.viewDevDetails.bind(this);
        this.closeDeleteModal = this.closeDeleteModal.bind(this);
        this.handleChange = this.handleChange.bind(this);
        this.filterChange = this.filterChange.bind(this);
        this.addFilter = this.addFilter.bind(this);
        this.removeFilter = this.removeFilter.bind(this);
        this.pagnCallback = this.pagnCallback.bind(this);
    }

    async componentDidMount() {
        this.setTables(); 
    }

    showAddModal(){
        this.setState({
            showEditModal: true,
            updateType: 'add',
            updateDeviceId: '',
        });
    }

    showFilterModal() {
        const filterTable = {head: ['Filter', 'Condition', 'Remove'], body: []};
        for (const key of Object.keys(this.state.filter.option)) {
            const row = [key, this.state.filter.option[key], <MdDelete onClick={() => this.removeFilter(key)} className="oprnButton"/>]
            filterTable.body.push(row);
        }
        const input = 
            <Row className="mb-3">
                <Col>
                    <select value={this.state.filterCurrent.select} onChange={this.filterChange} name="select" 
                        className="form-control"
                    >
                        <option hidden value="">Select Filter</option>
                        {this.state.filter.type.map((val) => (<option key={val} value={val}>
                            {val}
                        </option>))}
                    </select>
                </Col>
                <Col className="d-flex justify-content-between">
                    <Form.Control type="text" name="text" onChange={this.filterChange} 
                        className="mr-2" placeholder="Keyword" value={this.state.filterCurrent.text}
                    />
                    <h3><MdAddCircleOutline onClick={this.addFilter} className="oprnButton"/></h3>
                </Col>
            </Row>
        const showOprn = false; const alignLeft = false;
        const filter =
            <div>
                {(filterTable.body.length > 0) && 
                    <div className="mb-3">
                        <LiveDataTable tableBody={filterTable.body} tableHead={filterTable.head} showOprn={showOprn} 
                            alignLeft={alignLeft}
                        />
                    </div>
                }
                {input}
            </div>
        const modalHead = 
            <div className="d-flex justify-content-start">
                Apply Filter 
                <summary>
                    <h3>
                        <DarkOverlays hovertext="Remove Filter" placement="bottom">
                            <MdRemoveCircleOutline className="ml-2 oprnButton" onClick={() => this.setTables()}/>
                        </DarkOverlays>
                    </h3>
                </summary>
            </div>
        this.setState({
            showDeleteModal: true,
            modal: {head: modalHead, content: {body: filter, buttons: true,
                submitLink: 'filterDevice'}},
            apiCall: {restore: '', permanent: ''}
        });
    }

    addFilter() {
        if(this.state.filterCurrent.select && this.state.filterCurrent.text) {
            const filter = {...this.state.filter};
            const types = filter.type.filter(type => type !== this.state.filterCurrent.select);
            const option = {...filter.option, ...{[this.state.filterCurrent.select]: this.state.filterCurrent.text}}
            this.setState({filter: {type: types, option: option}, filterCurrent: {select: '', text: ''}}, 
                () => this.showFilterModal()); 
        } else {
            toastr.error('Oops !!', 'Select Both Filter and Keyword');
        }
    }

    removeFilter(select) {
        const filter = {...this.state.filter};
        const types = filter.type; types.push(select);
        const option = filter.option; delete option[select];
        this.setState({filter: {type: types, option: option}, filterCurrent: {select: '', text: ''}}, 
            () => this.showFilterModal());
    }

    filterChange(e) {
        if(e.target.name === 'select') {
            const text = this.state.filterCurrent.text;
            this.setState({filterCurrent: {select: e.target.value, text: text}}, () => this.showFilterModal());
        } else {
            const select = this.state.filterCurrent.select;
            this.setState({filterCurrent: {text: e.target.value, select: select}}, () => this.showFilterModal());
        }
    }

    showUpdateModal(id) {
        this.setState({
            showEditModal: true,
            updateType: 'update',
            updateDeviceId: id,
        });
    }

    closeEditModal() {
        this.setState({showEditModal: false});
    }

    showTempDeleteModal(id) {
        const input = 
            <Form.Control type="text" onChange={this.handleChange} className="w-100 mb-3" placeholder="Reason For Deactivation" />
        this.setState({
            showDeleteModal: true,
            deleteDeviceId: id,
            modal: {head: 'Continue Deactivating the Device ?', content: {body: input, buttons: true,
                submitLink: 'restoDltDevice'}},
            apiCall: {restore: false, permanent: false}
        });
    }

    showDeleteModal(id) {
        this.setState({
            showDeleteModal: true,
            deleteDeviceId: id,
            modal: {head: 'Continue Deleting the Device ?', content: {body: '', buttons: true,
                submitLink: 'restoDltDevice'}},
            apiCall: {restore: false, permanent: true}
        });
    }

    showRestoreModal(id) {
        this.setState({
            showDeleteModal: true,
            deleteDeviceId: id,
            modal: {head: 'Continue Restoring the Device ?', content: {body: '', buttons: true,
                submitLink: 'restoDltDevice'}},
            apiCall: {restore: true, permanent: ''}
        });
    }

    viewDevDetails(id) {
        const devDet = <div><pre>{JSON.stringify(this.props.deviceList[id], null, 4)}</pre></div>
        this.setState({
            showDeleteModal: true,
            modal: {head: 'Device: ' + id, content: {body: devDet, buttons: false,
                submitLink: ''}},
            apiCall: {restore: '', permanent: ''}
        });
    }

    closeDeleteModal() {
        this.setState({ showDeleteModal: false, filter: {type: ['deviceId', 'zone', 'city', 'subType'], option: {}},
            filterCurrent: {text: '', select: ''}});
    }

    async setTables() {
        const loader = <div className="text-center">
            <Loader type="MutatingDots" color="#00BFFF" height={100} width={100} />
            <br/><strong className="lead">Loading !!</strong></div>
        this.setState({table : {activated: loader , deactivated: loader, actCnt: 0, deactCnt: 0}});
        
        let extraOption = {};
        const filter = {...this.state.filter.option}; const array = ['deviceId', 'zone', 'city', 'subType'];
        array.forEach(filt => {
            if(filter.hasOwnProperty(String(filt))) {
                extraOption = {...extraOption, ...{[filt]: this.state.filter.option[filt]}}
            } else {
                extraOption = {...extraOption, ...{[filt]: 'null'}}
            }
        })

        const devCountAct = await this.props.getDeviceCount({...extraOption, ...{activated: true}});
        const devCountDeact = await this.props.getDeviceCount({...extraOption, ...{activated: false}});
        const showOprn = true; const alignLeft = true; let activated; let deactivated;
        const notFound = <div className="text-center">
            <img src={process.env.PUBLIC_URL + "/img/notFound.png"} alt="No Data"/>
            <br/><strong className="lead">No Data !!</strong></div>
        
        if(devCountAct > 0) {
            const tableHeadAct = ['Status', 'City', 'Device', 'Landmark', 'Operation'];
            if(this.props.hideOprn) {tableHeadAct.pop(); showOprn = false}
            const devicesAct = await this.props.getAllDevices({...extraOption,
                ...{limit: this.state.limit, offset: this.state.offset.act, activated: true}});
            const tableBodyAct = devicesAct.map(device => {
                if(this.props.enableSltDev && this.state.allowSelecting) {
                    this.props.updateSelectedDevice(device.deviceId);
                    this.setState({allowSelecting: false})
                }
                return(['Active',device.location.city, device.deviceId,device.location.landMark]);
            })
            activated = 
                <LiveDataTable tableBody={tableBodyAct} tableHead={tableHeadAct} showOprn={showOprn} 
                    onEdit={this.showUpdateModal} onDelete={this.showTempDeleteModal} alignLeft={alignLeft}
                    onView={this.viewDevDetails}
                />;
        } else {
            activated = notFound;
        }

        if(devCountDeact > 0) {
            const tableHeadDeact = ['City', 'Device', 'Landmark', 'Deact User', 'Deact Time', 'Reason', 'Operation'];
            const devicesDeact = await this.props.getAllDevices({...extraOption,
                ...{limit: this.state.limit, offset: this.state.offset.deact, activated: false}});
            const tableBodyDeact = devicesDeact.map(device => {
                const date = new Date(device.deactLog.date + ' UTC').toLocaleString("en-US");
                return([device.location.city, device.deviceId,device.location.landMark, device.deactLog.user,
                    date, device.deactLog.reason]);
            })
            const restore = true;
            deactivated = 
                <LiveDataTable tableBody={tableBodyDeact} tableHead={tableHeadDeact} showOprn={showOprn} 
                    onEdit={this.showUpdateModal} onDelete={this.showDeleteModal} alignLeft={alignLeft}
                    onView={this.viewDevDetails} onRestore={this.showRestoreModal} restore={restore}
                />;
        } else {
            deactivated = notFound;
        }

        this.setState({table : {activated: activated , deactivated: deactivated, actCnt: devCountAct, deactCnt: devCountDeact}}, 
            () => this.closeDeleteModal()) 
    }

    async restoDltDevice(e) {
        e.preventDefault();
        try {
            let status;
            if (this.state.apiCall.restore) {
                const devDetails = {...this.props.deviceList[this.state.deleteDeviceId]};
                devDetails.activated = true;
                devDetails.deactLog = "";
                status = await this.props.updateDeviceDetails(devDetails);
            } else {
                if(this.state.apiCall.permanent) {
                    status = await this.props.deleteDevice({deviceId: this.state.deleteDeviceId}); 
                } else {
                    const devDetails = {...this.props.deviceList[this.state.deleteDeviceId]};
                    devDetails.activated = false;
                    devDetails.deactLog = {
                        user: sessionStorage.getItem("whoami"),
                        date: new Date().toISOString().replace(/T/, ' ').replace(/\..+/, ''),
                        reason: this.state.deactivationReason
                    };
                    status = await this.props.updateDeviceDetails(devDetails)
                }
            }
            if (!status) {
                toastr.error('Oops !!', 'Some Error Occured');
            } else {
                const info = (this.state.apiCall.restore) ? 'Device Restored' : 
                    (this.state.apiCall.permanent) ? 'Device Deleted' : 'Device Deactivated';
                toastr.success('Success !!', info);
            }
        } catch(err) {
            toastr.error('Oops !!', 'Some Error Occured');
        }
        this.setTables(); this.closeDeleteModal(); 
    }

    handleChange(event) {
        this.setState({deactivationReason: event.target.value})
    }

    pagnCallback(type, newOffset) {
        const sibling = (type === 'act') ? 'deact' : 'act'
        this.setState({offset: {[type]: newOffset, [sibling]: this.state.offset[sibling]}}, 
            () => this.setTables())
    }

    render() {
        const disableSwitch = (this.state.filter.option && (Object.keys(this.state.filter.option).length > 0)) ? false : true;
        return (
            <div className="pt-3">
                <Container>
                    <div className="d-flex justify-content-between">
                        <h2>Devices</h2>
                        <div>
                            <h2 className="d-flex justify-content-start">
                                <DarkOverlays hovertext="Filter" placement="bottom">
                                    <summary>
                                        <FiFilter className="oprnButton" onClick={this.showFilterModal}/>
                                    </summary>
                                </DarkOverlays>
                                <DarkOverlays hovertext="Add" placement="bottom">
                                    <summary>
                                        <MdAddCircleOutline className="oprnButton" onClick={this.showAddModal}/>
                                    </summary>
                                </DarkOverlays>
                            </h2>
                        </div>
                    </div>
                    {this.state.table.activated}
                    <Paginations userCount={this.state.table.actCnt} limit={this.state.limit}
                        offset={this.state.offset.act} pagnCallback={this.pagnCallback} type="act"
                    />
                    <h2 className="mt-5">Deactivated Devices</h2>
                    {this.state.table.deactivated}
                    <Paginations userCount={this.state.table.deactCnt} limit={this.state.limit}
                        offset={this.state.offset.deact} pagnCallback={this.pagnCallback} type="deact"
                    />
                </Container>
                {this.state.showEditModal && 
                    <DeviceEditForm show={this.state.showEditModal} handleClose={this.closeEditModal}
                        updateDeviceId={this.state.updateDeviceId} updateType={this.state.updateType}
                        updateTable={() => this.setTables()}
                    /> 
                }
                <Modal show={this.state.showDeleteModal} onHide={this.closeDeleteModal}
                    aria-labelledby="contained-modal-title-vcenter" scrollable="true"
                >
                    <Modal.Header closeButton>
                        <Modal.Title>{this.state.modal.head}</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        {this.state.modal.content.body}
                        {this.state.modal.content.buttons &&
                            <Row>
                                <Col sm={6}>
                                    <Button className="w-100" variant="secondary" onClick={this.closeDeleteModal}>
                                        Cancel
                                    </Button>
                                </Col>
                                <Col sm={6}>
                                    {(this.state.modal.content.submitLink === 'restoDltDevice') &&
                                        <Button className="w-100" onClick={(e) => this.restoDltDevice(e)}>
                                            Yes
                                        </Button>
                                    }
                                    {(this.state.modal.content.submitLink === 'filterDevice') &&
                                        <Button disabled={disableSwitch} className="w-100" onClick={() => this.setTables()}>
                                            Filter
                                        </Button>
                                    }
                                </Col>
                            </Row>
                        }
                    </Modal.Body>
                </Modal>
            </div>
        )
    }
}

function mapStateToProps(state) {
    return ({
        deviceList: state.devices.data
    });
}

export function mapDispatchToProps(dispatch) {
    return ({
        deleteDevice: (data) => {
            return dispatch(deleteDevice(dispatch, data))
        },
        getDeviceCount: (options) => {
            return dispatch(getDeviceCount(dispatch, options))
        },
        getAllDevices: (options) => {
            return dispatch(getAllDevices(dispatch, options))
        },
        updateDeviceDetails: (data) => {
            return dispatch(updateDeviceData(dispatch, data))
        },
    });
}

export default connect(mapStateToProps, mapDispatchToProps)(DeviceTable);
