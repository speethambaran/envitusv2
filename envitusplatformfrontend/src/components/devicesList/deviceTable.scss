@import '../../styles/index.scss';

.oprnButton:hover {
    color: $base-hover-color;
}
