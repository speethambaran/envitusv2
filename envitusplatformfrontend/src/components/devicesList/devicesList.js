import React from 'react';
import { connect } from 'react-redux';
import Sensor from '../sensor/sensor';
import { updateSelectedDevice } from '../../action/deviceAction';
import { getDeviceCount, getDeviceDetails} from '../../services/deviceApi';
import { Table } from 'react-bootstrap';
import { MdModeEdit, MdDelete } from "react-icons/md";

export class DeviceList extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            deviceList: null,
        };
    }

    async componentDidMount() {
        const option = {deviceId: 'null', zone: 'null', city: 'null', subType: 'null', activated: 'null'}
        const count = await this.props.getDeviceCount(option);
        (this.props.listType === "sensor") ? await this.formatSensorList(count) : (
            (this.props.listType === "sensor3D") ? await this.formatSensor3DList(count) 
                : await this.formatTableList(count)
        );	
    }

    async componentDidUpdate(prevProps) {
        if (prevProps.updateList !== this.props.updateList || prevProps.selectedDevice!== this.props.selectedDevice) {
            const option = {deviceId: 'null', zone: 'null', city: 'null', subType: 'null', activated: 'null'}
            const count = await this.props.getDeviceCount(option);
            (this.props.listType === "sensor") ? await this.formatSensorList(count) : (
                (this.props.listType === "sensor3D") ? await this.formatSensor3DList(count) 
                    : await this.formatTableList(count)
            );	
        }
    }

    selectSensor(id) {
        if(!this.props.showDeviceOprn && this.props.selectedDevice !== id) {
            this.props.updateSelectedDevice(id);
        }
    }

    async formatSensorList(count) {
        const sensorElements = [];
        for (let i = 0; i < count; i++) {
            const sensorData = await this.props.getDeviceDetails(i);
            const divStyle = {
                top: sensorData.location.latitude + '%',
                left: sensorData.location.longitude + '%',
            };
        
            sensorElements.push(
                <Sensor data-test="deviceListSensor" sensorId={sensorData.deviceId} 
                    key={sensorData.deviceId} mode="circle" positionStyle={divStyle} 
                />
            );
            
            if(i === 0 && !this.props.selectedDevice) {
                this.props.updateSelectedDevice(sensorData.deviceId);
            }			
        }
        this.setState({deviceList: <div>{sensorElements}</div>});
    }

    async formatSensor3DList(count) {
        const sensorElements = [];
        for (let i = 0; i < count; i++) {
            const sensorData = await this.props.getDeviceDetails(i);
            const divStyle = {
                slot: sensorData.location.slot,
                position: sensorData.location.dataPosition,
                normal: sensorData.location.dataNormal
            };

            sensorElements.push(
                <Sensor sensorId={sensorData.deviceId} key={sensorData.deviceId} mode="circle3D"
                    positionStyle={divStyle} data-test="deviceListSensor3D"
                />
            );
            
            if(i === 0 && !this.props.selectedDevice) {
                this.props.updateSelectedDevice(sensorData.deviceId);
            }			
        }
        this.setState({deviceList: [sensorElements]});
    }

    async formatTableList(count) {
        const sensorElements = [];
        for (let i = 0; i < count; i++) {
            const sensorData = await this.props.getDeviceDetails(i);
            sensorElements.push(
                <tr key={sensorData.deviceId} 
                    onClick={() => this.selectSensor(sensorData.deviceId)}
                    className={(this.props.selectedDevice === sensorData.deviceId) ? "selected-row" : ""}
                >
                    <td>{sensorData.deviceId}</td>
                    <td>{sensorData.type}</td>
                    <td>{sensorData.devFamily}</td>
                    <td>{sensorData.subType}</td>
                    {this.props.showDeviceOprn &&
                    <td className="editClass" data-test="formatOperation">
                        <MdModeEdit onClick={() => this.props.updateDevice(sensorData.deviceId)} 
                            data-test="operationEdit"
                        />
                        <MdDelete onClick={() => this.props.deleteDevice(sensorData.deviceId)}
                            data-test="operationDlt"
                        />
                    </td>
                    }
                </tr>
            );	
            
            if(!this.props.showDeviceOprn && i === 0 && !this.props.selectedDevice) {
                this.props.updateSelectedDevice(sensorData.deviceId);
            }
        }
        this.setState({deviceList: this.getTableStruture(sensorElements)});
    }

    getTableStruture(deviceRows) {
        return (
            <Table data-test="deviceListTable" striped bordered hover size="sm">
                <thead>
                    <tr>
                        <th>Device Id</th>
                        <th>Type</th>
                        <th>Family</th>
                        <th>Sub-Type</th>
                        {this.props.showDeviceOprn &&
                            <th data-test="operation">Operations</th>
                        }
                    </tr>
                </thead>
                <tbody>
                    {deviceRows}
                </tbody>
            </Table>
        );
    }
  
    render() {
        return this.state.deviceList;
    }
}

export function mapStateToProps(state) {
    return ({
        deviceCount: state.devices.deviceCount,
        selectedDevice: state.devices.selectedDevice
    });
}

export function mapDispatchToProps(dispatch) {
    return ({
        getDeviceCount: (option) => {
            return dispatch(getDeviceCount(dispatch, option))
        },
        getDeviceDetails: (id) => {
            return dispatch(getDeviceDetails(dispatch, id))
        },
        updateSelectedDevice: (id) => {
            return dispatch(updateSelectedDevice(id))
        }    
    });
}

export default connect(mapStateToProps, mapDispatchToProps)(DeviceList);
