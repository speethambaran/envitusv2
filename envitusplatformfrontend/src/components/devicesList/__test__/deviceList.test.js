import React from 'react';
import { shallow } from 'enzyme';   
import { DeviceList, mapStateToProps, mapDispatchToProps } from './../devicesList';
import { findByDataTest } from './../../../Utils';

const preConfig = (props={}) => {
    const component = shallow(<DeviceList {...props} />);
    return component;
};

describe('DeviceList Test', () => {
    describe('Snap Test', () => {
        it('match snapshot', () => {
            const wrapper = shallow(<DeviceList/>);   
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('State Dispach Prop', () => {
        it('should show previous device count value', () => {
            const initialState = {
                devices: {
                    deviceCount: 1
                }
            };
            expect(mapStateToProps(initialState).deviceCount).toBe(1);
        });

        it('should update selected device', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).updateSelectedDevice();
            expect(dispatch.mock.calls[0][0]).toEqual({ type: 'UPDATE_SELECTED_DEVICE'});
        });

        it('getDeviceCount returns function', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).getDeviceCount();
            expect(typeof dispatch.mock.calls[0][0]).toBe('function');
        });

        it('getDeviceDetails returns function', () => {
            const dispatch = jest.fn();
            mapDispatchToProps(dispatch).getDeviceDetails();
            expect(typeof dispatch.mock.calls[0][0]).toBe('function');
        });
    });

    describe('function test', () => {
        describe('Component Mount', () => {
            it("component mount : listType - sensor: function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "sensor"
                };
                
                component = preConfig(props);
                component.instance().formatSensorList = jest.fn();
    
                return(component.instance().componentDidMount()).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatSensorList).toBeCalled();
                });            
            });
    
            it("component mount : listType - table: function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "table"
                };
    
                component = preConfig(props);
                component.instance().formatTableList = jest.fn();
    
                return(component.instance().componentDidMount()).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatTableList).toBeCalled();
                });
            });

            it("component mount : listType - sensor3D: function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "sensor3D"
                };
    
                component = preConfig(props);
                component.instance().formatSensor3DList = jest.fn();
    
                return(component.instance().componentDidMount()).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatSensor3DList).toBeCalled();
                });
            });
        })

        describe('Component Update', () => {
            it("component update : listType - sensor : propUpdateList != prevPropsUpdateList : function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "sensor",
                    updateList: "List1"
                };
    
                const mockPrevProps = {
                    updateList: "List2"  
                }
    
                component = preConfig(props);
                component.instance().formatSensorList = jest.fn();
    
                return(component.instance().componentDidUpdate(mockPrevProps)).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatSensorList).toBeCalled();
                });
            });
    
            it("component update : listType - table : propUpdateList != prevPropsUpdateList : function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "table",
                    updateList: "List1"
                };
    
                const mockPrevProps = {
                    updateList: "List2"  
                }
    
                component = preConfig(props);
                component.instance().formatTableList = jest.fn();
    
                return(component.instance().componentDidUpdate(mockPrevProps)).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatTableList).toBeCalled();
                });
            });
            
            it("component update : listType - sensor3D : propUpdateList != prevPropsUpdateList : function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "sensor3D",
                    updateList: "List1"
                };
    
                const mockPrevProps = {
                    updateList: "List2"  
                }
    
                component = preConfig(props);
                component.instance().formatSensor3DList = jest.fn();
    
                return(component.instance().componentDidUpdate(mockPrevProps)).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(1);
                    expect(component.instance().formatSensor3DList).toBeCalled();
                });
            });
    
            it("component update : listType - any : propUpdateList = prevPropsUpdateList : function calling ?", async() => {
                let component;
                const mockgetDeviceCount = jest.fn();
                
                const props = {
                    getDeviceCount: mockgetDeviceCount,
                    listType: "sensor",
                    updateList: "List1"
                };
    
                const mockPrevProps = {
                    updateList: "List1"  
                }
    
                component = preConfig(props);
                component.instance().formatSensorList = jest.fn();
    
                return(component.instance().componentDidUpdate(mockPrevProps)).then(() => {
                    expect(mockgetDeviceCount.mock.calls.length).toBe(0);
                    expect(component.instance().formatSensorList).toHaveBeenCalledTimes(0);
                });
            });
        })
        
        describe('formatSensorList', () => {
            it("formatSensorList : function calling ?", async () => {
                let component;
                const mockGetDevDetails = jest.fn();
                const mockUpdateSelectedDevice = jest.fn();
    
                mockGetDevDetails.mockReturnValueOnce({deviceId:'id1', type:'type1', devFamily:'family1', 
                subType: 'sub1', location: {latitude: 1, longitude: 2, slot: 'slot1', dataPosition: 'pos1', 
                dataNormal: 'nor1'}}).mockReturnValueOnce({deviceId:'id2',type:'type2', devFamily:'family2',
                subType: 'sub2', location: {latitude: 1, longitude: 2, slot: 'slot2', dataPosition: 'pos2', 
                dataNormal: 'nor2'}});      
    
                const props = {
                    getDeviceDetails: mockGetDevDetails,
                    updateSelectedDevice: mockUpdateSelectedDevice
                };
    
                component = preConfig(props);
    
                return(component.instance().formatSensorList(2)).then(() => {
                    expect(mockGetDevDetails.mock.calls.length).toBe(2);
                    const wrapperSensor = component.state("deviceList");
                    expect(wrapperSensor.props.children[0].props['data-test']).toBe('deviceListSensor');
                    expect(mockUpdateSelectedDevice.mock.calls.length).toBe(1);
                });
            });
        })

        describe('formatSensor3DList', () => {
            it("formatSensorList : function calling ?", async () => {
                let component;
                const mockGetDevDetails = jest.fn();
                const mockUpdateSelectedDevice = jest.fn();
    
                mockGetDevDetails.mockReturnValueOnce({deviceId:'id1', type:'type1', devFamily:'family1', 
                subType: 'sub1', location: {latitude: 1, longitude: 2, slot: 'slot1', dataPosition: 'pos1', 
                dataNormal: 'nor1'}}).mockReturnValueOnce({deviceId:'id2',type:'type2', devFamily:'family2',
                subType: 'sub2', location: {latitude: 1, longitude: 2, slot: 'slot2', dataPosition: 'pos2', 
                dataNormal: 'nor2'}});        
    
                const props = {
                    getDeviceDetails: mockGetDevDetails,
                    updateSelectedDevice: mockUpdateSelectedDevice
                };
    
                component = preConfig(props);
    
                return(component.instance().formatSensor3DList(2)).then(() => {
                    expect(mockGetDevDetails.mock.calls.length).toBe(2);
                    const wrapperSensor = component.state("deviceList");
                    expect(wrapperSensor[0][0].props['data-test']).toBe('deviceListSensor3D');
                    expect(mockUpdateSelectedDevice.mock.calls.length).toBe(1);
                });
            });
        })
        
        describe('formatTableList', () => {
            it("formatTableList showOperation -> true: function calling ?", async () => {
                let component;
                const mockGetDevDetails = jest.fn();
                const mockUpdateDevice = jest.fn();
                const mockDeleteDevice = jest.fn();
                const mockUpdateSelectedDevice = jest.fn();
    
                mockGetDevDetails.mockReturnValueOnce(
                    {deviceId:'id', type:'type', devFamily:'family', subType: 'sub'}
                );      
    
                const props = {
                    showDeviceOprn: true,
                    getDeviceDetails: mockGetDevDetails,
                    updateDevice: mockUpdateDevice,
                    deleteDevice: mockDeleteDevice,
                    updateSelectedDevice: mockUpdateSelectedDevice
                };
    
                component = preConfig(props);
                component.instance().formatTableList(1);
                
                expect(mockGetDevDetails.mock.calls.length).toBe(1);
    
                return new Promise(resolve => setImmediate(resolve)).then(() => {
                    const wrapperOperation = findByDataTest(component, 'formatOperation');
                    expect(wrapperOperation.length).toBe(1);
                    
                    const wrapperEdit = findByDataTest(component, 'operationEdit');
                    wrapperEdit.simulate('click');
                    expect(mockUpdateDevice.mock.calls.length).toBe(1);
    
                    const wrapperDlt = findByDataTest(component, 'operationDlt');
                    wrapperDlt.simulate('click');
                    expect(mockDeleteDevice.mock.calls.length).toBe(1);
                    
                    expect(mockUpdateSelectedDevice.mock.calls.length).toBe(1);
                });
            });
    
            it("formatTableList showOperation -> false: function calling ?", async () => {
                let component;
                const mockGetDevDetails = jest.fn();
                const mockUpdateSelectedDevice = jest.fn();
    
                mockGetDevDetails.mockReturnValueOnce({deviceId:'id1', type:'type1', devFamily:'family1', 
                    subType: 'sub1'}).mockReturnValueOnce({deviceId:'id2', type:'type2', devFamily:'family2', 
                    subType: 'sub2'});      
    
                const props = {
                    showDeviceOprn: false,
                    getDeviceDetails: mockGetDevDetails,
                    updateSelectedDevice: mockUpdateSelectedDevice
                };
    
                component = preConfig(props);
                
                return(component.instance().formatTableList(2)).then(() => {
                    expect(mockGetDevDetails.mock.calls.length).toBe(2);
                    const wrapperOperation = findByDataTest(component, 'sensorTest');
                    expect(wrapperOperation.length).toBe(2);
                    expect(mockUpdateSelectedDevice.mock.calls.length).toBe(1);
                });
            });
        })
        
        describe('getTableStructure', () => {
            it("table Structure shows operation on prop show DeviceOprn boolean true", async () => {
                const props = {
                    showDeviceOprn: true,
                };
    
                const component = shallow(<DeviceList {...props} />);
    
                expect(component.instance().getTableStruture().props.children[0].props.children.props
                    .children[4].props.children).toBe('Operations');
            });
    
            it("table Structure shows device select on prop show DeviceOprn boolean false", async () => {
                const props = {
                    showDeviceOprn: false,
                };
    
                const component = shallow(<DeviceList {...props} />);
    
                expect(component.instance().getTableStruture().props.children[0].props.children.props
                    .children[4]).toBe(false);
            });  
        })
    });
});