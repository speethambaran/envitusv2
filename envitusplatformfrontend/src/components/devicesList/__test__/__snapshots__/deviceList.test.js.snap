// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceList Test Snap Test match snapshot 1`] = `""`;
